src/gadgets/point_doubling/double.rs

```diff
@@ -0,0 +1,194 @@
+pub fn point_doubling_proof(
+    pc_gens: &PedersenGens,
+    bp_gens: &BulletproofGens,
+    p1: RistrettoPoint,
+    p2: RistrettoPoint,
+    a: FieldElement,
+    d: FieldElement,
+) -> Result<R1CSProof, R1CSError> {
+    let mut transcript = Transcript::new(b"R1CS Point Add Gadget");
+
+    // Create the prover->
+    let mut prover = Prover::new(pc_gens, &mut transcript);
+
+    // Commit high-level variables
+    // Get LCs for P1 and 2*P1 => P2
+    let mut lcs = n_point_coords_to_LC(&[p1, p2]);
+    // Get a and d as LC
+    lcs.push((
+        fq_as_scalar(a).into(),
+        fq_as_scalar(d).into(),
+        fq_as_scalar(d).into(),
+        fq_as_scalar(d).into(),
+    ));
+
+    // Build the CS
+    // XXX: We should get the z and t and verify that it satisfies the curve eq
+    // in another gadget.
+    let (x, y, _, _) = point_doubling_gadget(
+        &mut prover,
+        lcs[0].clone(),
+        lcs[2].1.clone(),
+        lcs[2].0.clone(),
+    );
+    point_doubling_constrain_gadget(&mut prover, lcs[1].clone(), (x.into(), y.into()));
+
+    // Build the proof
+    let proof = prover.prove(bp_gens)?;
+    Ok(proof)
+}
+
+/// Verifies a proof which holds the constraints related to
+/// the point doubling of a publicly known RistrettoPoint.
+pub fn point_doubling_verify(
+    pc_gens: &PedersenGens,
+    bp_gens: &BulletproofGens,
+    p1: RistrettoPoint,
+    p2: RistrettoPoint,
+    a: FieldElement,
+    d: FieldElement,
+    proof: R1CSProof,
+) -> Result<(), R1CSError> {
+    let mut transcript = Transcript::new(b"R1CS Point Add Gadget");
+
+    // Create the verifier
+    let mut verifier = Verifier::new(&mut transcript);
+
+    // Commit high-level variables
+    // Get LCs for P1 and 2*P1 => P2
+    let mut lcs = n_point_coords_to_LC(&[p1, p2]);
+    // Get a and d as LC
+    lcs.push((
+        fq_as_scalar(a).into(),
+        fq_as_scalar(d).into(),
+        fq_as_scalar(d).into(),
+        fq_as_scalar(d).into(),
+    ));
+
+    // Build the CS
+    // XXX: We should get the z and t and verify that it satisfies the curve eq
+    // in another gadget.
+    let (x, y, _, _) = point_doubling_gadget(
+        &mut verifier,
+        lcs[0].clone(),
+        lcs[2].1.clone(),
+        lcs[2].0.clone(),
+    );
+    point_doubling_constrain_gadget(&mut verifier, lcs[1].clone(), (x.into(), y.into()));
+
+    verifier.verify(&proof, &pc_gens, &bp_gens, &mut thread_rng())
+}
+
+/// Builds and adds to the CS the circuit that corresponds to the
+/// doubling of a Twisted Edwards point in Extended Coordinates.
+pub fn point_doubling_gadget(
+    cs: &mut ConstraintSystem,
+    (p1_x, p1_y, p1_z, p1_t): (LC, LC, LC, LC),
+    d: LC,
+    a: LC,
+) -> (Variable, Variable, Variable, Variable) {
+    // Point doubling impl
+    // A = p1_x²
+    // B = p1_y²
+    // C = 2*p1_z²
+    // D = a*A
+    // E = (p1_x + p1_y)² - A - B
+    // G = D + B
+    // F = G - C
+    // H = D - B
+    // X3 = E * F,  Y3 = G * H, Z3 = F * G, T3 = E * H
```

Comment:
Maybe just write it `X3 = E * F,  Y3 = G * H, T3 = E * H Z3 = F * G` so its inline with the usual format.

Reply:
The zerocaf's RistrettoPoint follows the same structure and order as the `curve25519-dalek` RistrettoPoint which is (X, Y, Z, T). That's why it is written like this.

Reply:
Fair enough, not a big deal, the fact that T is there is showing its extended. Its just different in the [paper](https://eprint.iacr.org/2008/522.pdf)